### PR TITLE
Event listener overhead reduction pt. 3

### PR DESCRIPTION
**Pull request by MihaZupan**
Building on top of #51822 and #52092.

- Reduced the size of `EventWrittenEventArgs` for the common case

    - As a minimum, events need the `EventSource`, `EventId`, `TimeStamp` and `Payload`.
    - Moved the rest to a `MoreInfo` class that's allocated only if more parameters are set:
    - ActivityId, RelatedActivityId, Tags, Opcode, Level, Keywords, ThreadId, PayloadNames, EventName, Message
- Stored `ParameterNames` on `EventMetadata` instead of `EventWrittenEventArgs`
- Avoided duplication around `TraceLoggingEventTypes` 293538e

[Benchmarks.cs](https://gist.github.com/MihaZupan/36432bc09090485595063caef4ae0929)
|                Method |                  Toolchain |        Mean |     Error | Ratio |  Gen 0 | Allocated |
|---------------------- |--------------------------- |------------:|----------:|------:|-------:|----------:|
|    WriteEventNoParams |                       base |    74.98 ns |  0.635 ns |  1.00 | 0.0440 |     184 B |
|    WriteEventNoParams | after 52092 (current main) |    56.56 ns |  0.447 ns |  0.75 | 0.0325 |     136 B |
|    WriteEventNoParams |                        new |    49.36 ns |  0.430 ns |  0.66 | 0.0134 |      56 B |
|                       |                            |             |           |       |        |           |
| WriteHttpRequestStart |                       base |   809.63 ns |  5.650 ns |  1.00 | 0.1068 |     448 B |
| WriteHttpRequestStart | after 52092 (current main) |   206.04 ns |  2.987 ns |  0.25 | 0.1070 |     448 B |
| WriteHttpRequestStart |                        new |   186.86 ns |  1.385 ns |  0.23 | 0.0880 |     368 B |
|                       |                            |             |           |       |        |           |
|       FullYarpRequest |                       base | 5,913.87 ns | 43.604 ns |  1.00 | 1.2054 |    5048 B |
|       FullYarpRequest | after 52092 (current main) | 1,973.14 ns |  7.223 ns |  0.33 | 1.1368 |    4761 B |
|       FullYarpRequest |                        new | 1,764.72 ns | 10.423 ns |  0.30 | 0.7744 |    3241 B |

Should be all from me around EventSource for a while.

PR adding a subset of these benchmarks to the performance repo: https://github.com/dotnet/performance/pull/1800

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 1]
```diff
@@ -4830,46 +4724,63 @@ public long OSThreadId
...
-                    m_osThreadId = (long)Thread.CurrentOSThreadId;
+                    moreInfo.OsThreadId = (long)Thread.CurrentOSThreadId;
 #endif
                 }
 
-                return m_osThreadId.Value;
+                return moreInfo.OsThreadId.Value;
             }
-            internal set => m_osThreadId = value;
+            internal init => MoreInfo.OsThreadId = value;
```
Comment (MihaZupan) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:4739]: `OSThreadId` has a similar issue as `ActivityId` (#42128) where the returned value depends on the caller's thread.

The difference is that `OSThreadId` saves the first caller's value where as `ActivityId` does not. Should these be changed to have the same behavior?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 2]
```diff
@@ -2489,18 +2436,69 @@ internal partial struct EventMetadata
...
 
-            public TraceLoggingEventTypes? TraceLoggingEventTypes;
             public EventActivityOptions ActivityOptions;
+
+            private TraceLoggingEventTypes _traceLoggingEventTypes;
+            public TraceLoggingEventTypes TraceLoggingEventTypes
+            {
+#if !ES_BUILD_STANDALONE
+                [RequiresUnreferencedCode(EventSourceRequiresUnreferenceMessage)]
+#endif
```
Comment (MihaZupan) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2453]: @LakshanF is adding this attribute sufficient here?

**Comment by noahfalk** [comment]
@josalem @davmason @sywhang

**Review by noahfalk — APPROVED** [review]
This looked good to me though I'd like one off @josalem, @sywhang, @davmason to take a look as well.

**Review by josalem — APPROVED** [review]
I read this one when it first went up, but apparently forgot to hit the approve button!  Thanks!

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 3]
```diff
@@ -4636,25 +4625,19 @@ public Guid ActivityId
         {
             get
             {
-                Guid activityId = m_activityId;
-                if (activityId == Guid.Empty)
+                if (_moreInfo is MoreEventInfo moreInfo && moreInfo.ActivityId != Guid.Empty)
```
[the remark this answers is not in the document]
Comment (noahfalk) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:4628]: My guess is that the usage would be uncorrelated, but if both are independently common enough then there could still be  significant overlap - this is just educated guessing on my part. If the perf consequence isn't that meaningful then there is no need to worry, but I assumed you wouldn't have done the optimization at all if it didn't make a difference.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs (hunk 1 above)
Comment (MihaZupan) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs]: Changed

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MihaZupan) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs]: |                        Method |                Toolchain |        Mean |     Error | Ratio |  Gen 0 | Allocated |
|------------------------------ |------------------------- |------------:|----------:|------:|-------:|----------:|
|            WriteEventNoParams |  OnExtraInfo\corerun.exe |    55.46 ns |  1.125 ns |  1.00 | 0.0134 |      56 B |
|            WriteEventNoParams | OnBaseObject\corerun.exe |    53.52 ns |  0.420 ns |  0.96 | 0.0172 |      72 B |
|                               |                          |             |           |       |        |           |
| WriteEventNoParams_ActivityId |  OnExtraInfo\corerun.exe |    84.18 ns |  0.631 ns |  1.00 | 0.0401 |     168 B |
| WriteEventNoParams_ActivityId | OnBaseObject\corerun.exe |    77.48 ns |  0.679 ns |  0.92 | 0.0172 |      72 B |
|                               |                          |             |           |       |        |           |
|               FullYarpRequest |  OnExtraInfo\corerun.exe | 1,939.21 ns | 38.206 ns |  1.00 | 0.7744 |    3241 B |
|               FullYarpRequest | OnBaseObject\corerun.exe | 1,948.20 ns | 20.337 ns |  1.00 | 0.8469 |    3545 B |
|                               |                          |             |           |       |        |           |
|    FullYarpRequest_ActivityId |  OnExtraInfo\corerun.exe | 5,101.87 ns | 85.053 ns |  1.00 | 2.4490 |   10266 B |
|    FullYarpRequest_ActivityId | OnBaseObject\corerun.exe | 4,558.92 ns | 31.574 ns |  0.89 | 1.6022 |    6713 B |

Moved it back onto the base object.
Looks like a measurable improvement for the case when ActivityId is accessed while not really affecting other cases.

**Review by sywhang — APPROVED** [review]
LGTM

**Comment by MihaZupan** [comment]
Failure: #11063

**Pull request merged**